File: src/main/java/com/mercateo/jsonschema/property/MethodCollector.java
package com.mercateo.jsonschema.property;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.lang.reflect.Type;

import com.mercateo.jsonschema.generictype.GenericType;
import javaslang.collection.Stream;

public final class MethodCollector implements RawPropertyCollector {

    private final AnnotationMapBuilder annotationMapBuilder;

    public MethodCollector() {
        this.annotationMapBuilder = new AnnotationMapBuilder();
    }

    @Override
    public Stream<RawProperty> forType(GenericType<?> genericType) {
        return Stream.of(genericType.getRawType().getDeclaredMethods())
                .filter(method -> !method.isSynthetic())
                .filter(method -> method.getDeclaringClass() != Object.class)
                .filter( method -> method.getReturnType() != void.class)
                .filter(method -> method .getParameterCount() == 0)
                .map(method -> mapRawDataProperty(method, genericType.getType()));
    }

    private RawProperty mapRawDataProperty(Method method, Type type) {
        final String methodName = getPropertyName(method);
        return ImmutableRawProperty.of(methodName, GenericType.of(method, type),
                annotationMapBuilder.createMap(method.getAnnotations()), (
                        Object object) -> valueAccessor(method, object));
    }

    private String getPropertyName(Method method) {
        final String methodName = method.getName();
        if (methodName.startsWith("get") && Character.isUpperCase(methodName.charAt(3))) {
            return Character.toLowerCase(methodName.charAt(3)) + methodName.substring(4);
        }
        if (methodName.startsWith("is") && Character.isUpperCase(methodName.charAt(2))) {
            return Character.toLowerCase(methodName.charAt(2)) + methodName.substring(3);
        }
        return methodName;
    }

    private Object valueAccessor(Method method, Object object) {
        try {
            return method.invoke(object);
        } catch (IllegalAccessException | InvocationTargetException e) {
            throw new IllegalStateException(e);
        }
    }
}

File: src/main/java/com/mercateo/jsonschema/property/MethodCollector.kt
package com.mercateo.jsonschema.property

import com.mercateo.jsonschema.generictype.GenericType
import java.lang.reflect.Method
import java.lang.reflect.Modifier

class MethodCollector(
        private val annotationMapBuilder: AnnotationMapBuilder = AnnotationMapBuilder()
) : RawPropertyCollector {

    override fun <S> forType(genericType: GenericType<S>): Sequence<RawProperty<S, *>> {
        return sequenceOf(*genericType.declaredMethods)
                .filter { !it.isSynthetic }
                .filter { it.declaringClass != Any::class.java }
                .filter { it.returnType != Void.TYPE }
                .filter { it.parameterCount == 0 }
                .filter { !Modifier.isStatic(it.modifiers) }
                .map { mapRawDataProperty(method = it, genericType = genericType) }
    }

    private fun <S> mapRawDataProperty(method: Method, genericType: GenericType<S>): RawProperty<S, *> {
        val methodType = GenericType.ofMethod(method, genericType.type) as GenericType<Any>
        return RawProperty<S, Any>(getPropertyName(method),
                methodType,
                annotationMapBuilder.createMap(*method.annotations),
                { instance: S -> valueAccessor(method, instance) })
    }

    private fun getPropertyName(method: Method): String {
        val methodName = method.name
        if (methodName.startsWith("get") && Character.isUpperCase(methodName[3])) {
            return Character.toLowerCase(methodName[3]) + methodName.substring(4)
        }
        if (methodName.startsWith("is") && Character.isUpperCase(methodName[2])) {
            return Character.toLowerCase(methodName[2]) + methodName.substring(3)
        }
        return methodName
    }

    private fun <S> valueAccessor(method: Method, instance: S): Any? {
        return method.invoke(instance)
    }
}
